sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/util/common/CounterNameAndMetadata.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.dataflow.sdk.util.common;
+
+import com.google.common.base.Strings;
+
+import javax.annotation.Nullable;
+
+/**
+ * A structured counter name.
+ *
+ * <p>It holds all information for generating structured name protos.
+ * It is also used as a wrapper during the migration to structured counters.
+ *
+ * <p>Before migration, the {@link CounterNameAndMetadata} will be converted to a flat name.
+ * During migration, the {@link CounterNameAndMetadata} will be mapped to a structured name proto.
+ */
+public class CounterNameAndMetadata {
+  /**
+   * Returns a CounterName with the given name.
+   */
+  public static CounterNameAndMetadata named(String name) {
+    return new CounterNameAndMetadata(name, "", "", "");
+  }
+
+  /**
+   * Returns a msecs CounterName.
+   */
+  public static CounterNameAndMetadata msecs(String name) {
+    return named(name + "-msecs");
+  }
+
+  /**
+   * Returns a CounterName identical to this, but with the given origin.
```

Comment:
`Returns a new {@link CounterNameAndMetadata} identical to this ...`

Also, should we just call this `CounterName`? Does the `AndMetadata` actually add any meaning beyond `CounterName`?